### DISCUSSION TITLE
Proposal: prop to remove detail panel icon

### DISCUSSION THREAD

**Discussion by peacechen** [material-table-core/core#87] [Ideas]
The detail panel icon takes up about 1/2 to 2/3 of a regular column width. For situations with a large number of columns, I'd like to add an option to remove that column entirely.
<img width="242" alt="Screen Shot 2021-01-25 at 1 45 34 PM" src="https://user-images.githubusercontent.com/6295083/105757478-9ee8b400-5f13-11eb-81af-5e8f9b6571c0.png">

I can set the icon to an empty component, but the column still exists and takes up space.

Additionally, I propose adding a prop to automatically expand the details panel on row click.  This can currently be done using the `togglePanel()` param in the onRowClick callback. It would be convenient for a prop to take care of that automatically.

@oze4 You had mentioned that there shouldn't be any new features added during the refactor. Is this a feature that I can add now, or will it have to wait?

**Comment by matthewoestreich** [comment]
I'm not sure I follow. You want to use the detail panel but have no column for the button (which allows you to open said detail panel)? How are you going to open it?

**Comment by matthewoestreich** [comment]
Ultimately, I would say to keep this in the back of your mind.

I wouldn't say this is a feature exactly, it's more of an abstraction to make it easier on the consumer. Due to the fact this doesn't resolve a critical bug, or add a feature to solve a critical issue, this will be put on the back burner for a little while. 

>Additionally, I propose adding a prop to automatically expand the details panel on row click. This can currently be done using the togglePanel() param in the onRowClick callback. It would be convenient for a prop to take care of that automatically.

Personally, I like the idea of "no magic" (or keeping things as un-opinionated as possible) internally, and allowing the user to accomplish things like this as you are currently accomplishing them.

**Reply by peacechen** [reply]
> I'm not sure I follow. You want to use the detail panel but have no column for for the button

Correct. The detail panel can be opened by clicking anywhere on the row. The second prop isn't as important. However the first one to hide the detail panel icon is needed because there's currently no way at all to prevent the column from rendering (when the detail panel is enabled, of course).

**Reply by matthewoestreich** [reply]
I see what you're saying, and I like the idea.

What about instead of adding a new prop (which I would consider to be the lazy fix), we change this...

https://github.com/material-table-core/core/blob/881a48a9d19d10b28fb3522a7f39e9a70fb7e5b7/src/components/m-table-body-row.js#L161-L166

...to something like this (if no icon is found it just returns)?

```js
const CustomIcon = ({ icon, iconProps }) => {
      if (!icon) {
        return;
      }
      if (typeof icon === 'string') {
        return <Icon {...iconProps}>{icon}</Icon>;
      }
      return React.createElement(icon, { ...iconProps });
    };
```

As far as the default row click expanding detail panel - that will have to be revisited. There's no real issue as a functioning solution exists.

**Reply by peacechen** [reply]
That could work. My (very minor) concern is that it overloads the meaning of that prop, and from that viewpoint it's the lazy fix :)

`m-table-header` will also need to be modified so that it doesn't render the header column.  The icon detection code may need to be in the parent so that it can be passed down to both the header and body row components.

**Reply by matthewoestreich** [reply]
I don't follow. Where is this overloading happening at? Also, one of the reasons for the refactor is too much prop drilling so this is something that could be part of a context. 

Good point ab the column, I'll look into that when I can.

**Reply by matthewoestreich** [reply]
Actually the real fix for this is a change in the API. which will occur after we get things stable. For now, if you use the newest version it shouldn't render the button

**Reply by peacechen** [reply]
Was a release made with the changes you mentioned above? I can hide the icon by passing in a blank element, but the column still renders. The goal is to eliminate that column to reclaim space for other data columns.

**Reply by matthewoestreich** [reply]
I thought I did but...yeah. I just published v2.3.23, which should have the updated code above which at least makes it easier to not show the icon.

It's not an issue with misunderstanding the goal, rather, manpower. I would rather focus my time working out the kinks and updating the codebase vs adding props/features like this that will be rewritten in a matter of weeks anyway.

Not that these issues are a waste of time, just that issues like this are not being prioritized (due to the fact this is more "sugar" than "resolving a critical bug"). There are hacks (whether css or what have you) that you could put in place on your end for the next couple of weeks. We apologize for the inconvenience.

[See here for our roadmap.](https://github.com/material-table-core/core/wiki/Roadmap)

In summary, all updates to v2 have stopped, outside of critical bugs.

As far as the feature request, it will have to be addressed at a later time (unless someone else wants to add it - for now I am unable to focus on things like this). Thank you so much for understanding!

**Reply by peacechen** [reply]
I totally understand the shortage of manpower. I'm happy to submit a PR with this change, but I need to know which branch to work off of. Every project has technical debt and there's usually a desire to rewrite/refactor. I worry that if this phase takes more than a few weeks, this project may lose favor as people peel off to other forks that are adding features.

**Reply by matthewoestreich** [reply]
The main branch. I'm not in this to generate interest, I'm in it to help make this better. If someone wants to use a diff library because we don't have X features until  couple weeks, they have every right to do so. I won't lose sleep over it.

**Reply by peacechen** [reply]
It's unlikely that mbrn will return to manage the original repo anytime soon. Even if he does, it's not sustainable for a widely used project to be hamstrung by one person.
I view this project's goal as helping active users who depend on it in their products. Part of that is long term refactors to sustain it, and another part is ongoing features and bugfixes. We need to build a community of contributors, and that will only happen when there are more people using this.

**Comment by peacechen** [comment]
With the delete icon (part of the editing features of the table), the detail panel icon real estate adds up. That's a good 1.5-2 columns worth of space.
<img width="640" alt="Screen Shot 2021-01-29 at 5 05 46 PM" src="https://user-images.githubusercontent.com/6295083/106336314-41769f00-6254-11eb-9412-fab482315afa.png">

When I get a chance to work on this, I'll make all the icons optional. If they're passed in as null, it won't render itself or the container & column it's in.

**Reply by matthewoestreich** [reply]
Yea I agree. How do you think we should fix it (what needs to be changed in the src code)? I'd be happy to merge any PR(s) for this!

**Comment by peacechen** [comment]
I went to implement this and noticed that you added the conditional to `m-table-body-row.js`:
https://github.com/material-table-core/core/blob/master/src/components/m-table-body-row.js#L161

Hopeful that your change would do the job, I first tried to pass in a falsy value for the DetailsPanel icon. Unfortunately, TS complains because the icons need to be wrapped in a forwardRef like so:
```
export const MaterialTableIconsWithoutDetail: Icons = {
  ...
  DetailPanel: forwardRef((props, ref) => null),
};
```

That prevents the icon from rendering, but the container button continues to render and take up space. We can't check for falsy `props.icons.DetailPanel` because it must be a function, even if that function returns null.

The only practical way to achieve this is to add a prop such as `options.showDetailPanelIcon`:
```
    if (!this.props.options.showDetailPanelIcon) {
      return null;
    }
```

I'll create a PR for this. `showDetailPanelIcon` will default to true.

**Reply by matthewoestreich** [reply]
Thank you! @Domino987 just wanted to get this on your radar. Cheers.

**Reply by peacechen** [reply]
Submitted PR #136

**Reply by Domino987** [reply]
Look good to me. Thank you.

**Comment by peacechen** [comment]
@oze4 @Domino987 
I submitted an additional PR #144 to fix the header row's alignment when `showDetailPanelIcon` is disabled. There are a few other PRs merged or in the pipeline that would valuable in the next release. When do you expect the next release will be?

**Comment by matthewoestreich** [comment]
#144 has been merged
